ko/pl.rst

```diff
@@ -0,0 +1,1383 @@
+        END LOOP;
+        CLOSE my_cursor;
+    END;
+
+커서는 위 예제처럼 명시적으로 OPEN, FETCH, CLOSE 실행문을 통해 이용할 수도 있고,
+이들 동작이 암묵적으로 이루어지는 For-Loop 문을 통해서 이용할 수도 있다.
+
+.. code-block:: sql
+
+    CREATE PROCEDURE test_proc(name VARCHAR, year INTEGER) AS
+
+        CURSOR my_cursor(a VARCHAR, y INTEGER) IS SELECT host_year, score FROM history
+            WHERE athlete = a and host_year >= y;
+    BEGIN
+
+        FOR r IN my_cursor(name, year) LOOP
+            put_line('host_year: ' || r.host_year || ' score: ' || r.score);
+        END LOOP;
+    END;
+
+
+내부 프로시저/함수 선언
+========================
+
+정의 중인 스토어드 프로시저/함수에서만 사용할 내부 프로시저/함수를 다음 문법에 따라 정의할 수 있다.
+어느 정도 규모를 이루거나 두 번 이상 반복되는 연관된 실행 과정을 내부 프로시저나 함수로 묶어 모듈화하면
+프로그램 가독성이 높아지고 유지 보수에 도움이 된다.
+
+::
+
+    <inner_procedure_decl> ::=
+        PROCEDURE <identifier> [ ( <seq_of_parameters> ) ] { IS | AS } [ <seq_of_declare_specs> ] <body> ;
+    <inner_function_decl> ::=
+        FUNCTION <identifier> [ ( <seq_of_parameters> ) ] RETURN <type_spec> { IS | AS } [ <seq_of_declare_specs> ] <body> ;
+
+    <seq_of_parameters> ::= <parameter> [, <parameter> ...]
+    <parameter> ::= <identifier> [ { IN | IN OUT | OUT } ] <type_spec>
+    <body> ::= BEGIN <seq_of_statements> [ EXCEPTION <seq_of_handlers> ] END [ <label_name> ]
+    <seq_of_declare_specs> ::= <declare_spec> [ <declare_spec> ... ]
+    <seq_of_statements> ::= <statement> ; [ <statement> ; ... ]
+    <seq_of_handlers> ::= <handler> [ <handler> ... ]
+    <handler> ::= WHEN <exception_name> [ OR <exeption_name> OR ... ] THEN <seq_of_statements>
+    <exception_name> ::= OTHER | identifier
+
+* *parameter*: 파라메터는 IN, IN OUT, OUT 세 가지 경우로 선언할 수 있다.
+* *type_spec*: :ref:`데이터 타입 <types>` 절에서 설명한 시스템 제공 타입
+* *body*: 필수적으로 하나 이상의 실행문과 선택적으로 몇 개의 Exception 핸들러로 구성된다.
+* *declare_spec*: 변수, 상수, Exception, 커서, Autonomous Transaction, 내부 프로시저, 내부 함수 선언 중 하나
+* *statement*: 아래 실행문 절 참조
+* *handler*: OR로 연결된 하나 이상의 Exception 이름들에 대하여 실행할 실행문들을 지정한다.
+* *exception_name*: OTHER인 경우 아직까지 매치되지 않은 모든 Exception에 매치된다. 아닌 경우는 시스템 Exception이거나 사용자 정의 Exception을 나타낸다.
+
+함수 *body*\에서는 RETURN 절에 지정된 타입에 맞는 값을 반환해야 한다.
+함수가 *body* 끝에 도달할 때까지 RETURN 문을 만나지 못하면 에러가 발생한다.
+프로시저의 RETURN 문에 반환값을 지정하면 에러이다.
+
+(내부) 프로시저/함수를 선언하면 자기 자신을 실행부에서 참조할 수 있다. 즉, 재귀호출이 가능하다.
+
+.. code-block:: sql
+
+    CREATE FUNCTION choose(m INT, n INT) RETURN INT AS
+
+        invalid_argument EXCEPTION;
+
+        -- 내부 함수 선언
+        FUNCTION factorial(n INT) RETURN INT AS
+        BEGIN
+            IF n < 0 THEN
+                RAISE invalid_argument;
+            ELSIF n <= 1 THEN
+                RETURN 1;
+            ELSE
+                RETURN n * factorial(n - 1);    -- 재귀 호출
+            end if;
+        END;
+    BEGIN
+        IF n > m OR n < 0 THEN
+            RAISE invalid_argument;
+        ELSE
+            RETURN factorial(m) / factorial(n) / factorial(m - n);
+        END IF;
+    END;
+
+동일한 선언부에서 선언된 내부 프로시저/함수끼리는 상호 재귀호출도 가능하다.
```

Comment:
When does the example below end?
Are there any errors during execution?

Reply:
It falls into an infinite loop and maybe ends on a kind of stack overflow exception. ^^
I'll add an end-condition.

Reply:
It meant that the manual needed a warning text.

Reply:
It meant that the manual needed a warning text.

Reply:
OK. I'll add a warning on infinite loops by mistakes